Agent/Services/ScriptingShellFactory.cs

```diff
@@ -0,0 +1,94 @@
+﻿using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.DependencyInjection;
+using Remotely.Shared.Enums;
+using System;
+using System.Threading.Tasks;
+
+namespace Remotely.Agent.Services;
+
+public interface IScriptingShellFactory
+{
+    Task<IExternalScriptingShell> GetOrCreateExternalShell(ScriptingShell shell, string senderConnectionId);
+    IPsCoreShell GetOrCreatePsCoreShell(string senderConnectionId);
+}
+
+internal class ScriptingShellFactory : IScriptingShellFactory
+{
+    private readonly MemoryCache _sessionCache = new(new MemoryCacheOptions());
+    private readonly IServiceProvider _serviceProvider;
+
+    public ScriptingShellFactory(IServiceProvider serviceProvider)
+    {
+        _serviceProvider = serviceProvider;
+    }
+    public IPsCoreShell GetOrCreatePsCoreShell(string senderConnectionId)
+    {
+        if (_sessionCache.TryGetValue(senderConnectionId, out var result) &&
+            result is IPsCoreShell shell &&
+            !shell.IsDisposed)
+        {
+            return shell;
+        }
+
+        shell = _serviceProvider.GetRequiredService<IPsCoreShell>();
```

Comment:
It looks like `IPsCoreShell` is registered transient but `PsCoreShell` takes in `IConfigService` as a dependency, which is registered as scoped. Should we change `IPsCoreShell` to be registered as scoped as well, and change this factory to take in a `IServiceScopeFactory` instead of `IServiceProvider`?

Reply:
We want `IPsCoreShell` to be Transient, because we want a new instance to be created every time we request it from DI.

We could technically make it scoped and create a new scope using `IServiceScopeFactory`, but then we'd need to manage the lifetimes of both the scope and the `IPsCoreShell`, since the resolved service is not supposed to outlive the scope.

However, scoped services are really meant for web applications, where the scope is automatically created per request.  All the scoped services here will function like singletons and should technically be registered as such.

I can do that now or in another PR.  Let me know.  :)

Reply:
I actually just added this to the branch I'm currently working on.